src/Microsoft.Health.Fhir.SqlServer/Features/Search/SqlServerSearchService.cs

```diff
@@ -90,13 +92,55 @@ internal class SqlServerSearchService : SearchService
 
         protected override async Task<SearchResult> SearchInternalAsync(SearchOptions searchOptions, CancellationToken cancellationToken)
         {
-            SearchResult searchResult;
+            SearchResult searchResult = await SearchImpl(searchOptions, SqlSearchType.Default, null, cancellationToken);
+            int resultCount = searchResult.Results.Count();
+            if (!searchOptions.IsSortWithFilter &&
+                searchResult.ContinuationToken == null &&
+                resultCount <= searchOptions.MaxItemCount &&
+                searchOptions.Sort != null &&
+                searchOptions.Sort.Count > 0 &&
+                searchOptions.Sort[0].searchParameterInfo.Code != KnownQueryParameterNames.LastUpdated)
+            {
+                // We seem to have run a sort which has returned less results than what max we can return.
+                // Let's determine whether we need to execute another query or not.
+                if ((searchOptions.Sort[0].sortOrder == SortOrder.Ascending && _didWeSearchForSortValue.HasValue && !_didWeSearchForSortValue.Value) ||
+                    (searchOptions.Sort[0].sortOrder == SortOrder.Descending && _didWeSearchForSortValue.HasValue && _didWeSearchForSortValue.Value))
+                {
+                    if (searchOptions.MaxItemCount - resultCount == 0)
+                    {
+                        // Since we are already returning MaxItemCount number of resources we don't want
+                        // to execute another search right now just to drop all the resources. We will return
+                        // a "special" ct so that we the subsequent request will be handled correctly.
+                        var ct = new ContinuationToken(new object[]
+                            {
+                                    "sentinelSortValue",
```

Comment:
nit: maybe use a different value?

Reply:
Is it possible to make this a constant and refer to the constant in both locations where the value is used?

Reply:
updated code to use a constant